File: solved/cf/Other/H_Farouk_and_Tape.py
def query(x):
    print(f"? {x}")
    return input()

def solve():
    start = -10**18
    target = query(start)
    step = 1
    while True:
        nxt = start + step
        if nxt > 10**18: nxt = 10**18
        if query(nxt) != target:
            low = start
            high = nxt
            break
        start = nxt
        step *= 2
        
    t1 = high
    while low <= high:
        mid = (low + high) // 2
        if query(mid) != target:
            t1 = mid
            high = mid - 1
        else:
            low = mid + 1
            
    color_t1 = query(t1)
    step = 1
    start = t1
    while True:
        nxt = start + step
        if nxt > 10**18: nxt = 10**18
        if query(nxt) != color_t1:
            low = start
            high = nxt
            break
        start = nxt
        step *= 2
        
    t2 = high
    while low <= high:
        mid = (low + high) // 2
        if query(mid) != color_t1:
            t2 = mid
            high = mid - 1
        else:
            low = mid + 1
            
    print(f"! {t2 - t1}")

for _ in range(int(input())) : 
    solve()

File: solved/cf/Other/H_Farouk_and_Tape.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long

string query(int x) {
    cout << "? " << x << endl;
    string res;
    if (!(cin >> res)) exit(0);
    return res;
}

void solve() {
    string color1 = query(0);
    string color2 = (color1 == "W") ? "B" : "W";
    int idx1 = 0, step = 1;
    while (true) {
        if (query(idx1 + step) == color2) break;
        idx1 += step;
        step *= 2;
    }
    int l = 0, r = step, ans = step;
    while (l <= r) {
        int mid = l + (r - l) / 2;
        if (query(idx1 + mid) == color2) {
            ans = mid;
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }
    int R = idx1 + ans;
    int idx2 = 0; 
    step = 1;
    while (true) {
        if (query(idx2 - step) == color2) break;
        idx2 -= step;
        step *= 2;
    }
    l = 0; 
    r = step;
    ans = step;
    while (l <= r) {
        int mid = l + (r - l) / 2;
        if (query(idx2 - mid) == color2) {
            ans = mid;
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }
    int L = idx2 - ans;

    cout << "! " << R - L - 1 << endl;
}

int32_t main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int t;
    if (cin >> t) {
        while (t--) solve();
    }
    return 0;
}